src/knightos/system.c

```diff
@@ -37,66 +37,44 @@ void ksleep(unsigned short msecs) __naked {
 
 void *malloc(size_t size) __naked {
 	__asm
+	POP HL
+	POP BC
+	PUSH BC
 	PUSH IX
-	LD IX, 0
-	ADD IX, SP
-
-	LD B, (IX + 5)
-	LD C, (IX + 4)
-
 	PCALL(MALLOC)
-	PUSH IX \ POP HL
-
-	JR Z, __malloc_done
-
-	LD HL, 0 ; NULL
-	POP IX
-	RET
-
-__malloc_done:
-	POP IX
-	RET
+	EX (SP), IX
+	EX (SP), HL
+	RET Z
+	LD HL, 0
```

Comment:
This is a __naked function and you have no RET at the end here. Also, while you're changing this, you might as well fill out errno.